Use JSDoc for generating docs

hekike opened the pull request:
## Issues

Please provide feedback in the comment section.

Closes:

* https://github.com/restify/node-restify/issues/1430

Use JSDoc for generating docs

# Changes

Required changes till phase one: automated docs for Server API from JSDoc

- [x] Add `make docs-build` to generate API docs from JSDoc
- [x] Add `make docs-watch` command to watch docs
- [x] Improve server JSDoc with examples
- [x] Improve JSDocs docs `server.js`: document events, exceptions and `pre`,` routed` and `after` states 
- [x] Improve audit plugin's JSDoc with examples
- [x] Improve jsonp plugin's JSDoc with examples
- [x] Split `lib/server.js` methods to separate functions for the sake of documentation
- [x] Reorder methods in server.js for cleaner and better docs / code

## Why documentation.js?

https://github.com/documentationjs/documentation

- works from JSDoc
- actively maintained
- markdown output
- watch capability for development
- Node API
- grouping feature
- can include external markdown files into the API documentation

Navigation in watch mode is not perfect, but for the content, it works well. 

## Migration plan

Replacing current API docs one by one see: `tools/docsBuild.js`.  
Start with ServerAPI docs.

**Proposed phases:**

1. **Server API docs *(This PR)***
  1.1 Move examples to JSDoc in `server.js`
  1.2 Improve JSDoc in `server.js`
  1.3 Add docs generating tool
2. Request Docs
3. Response Docs
4. Plugins Docs
5. Automatically generate docs at release

hekike commented:
@retrohacker I cleaned up the commit history and moved changes to separate commits. I recommend reviewing this PR one by one. 

Even if we would like to go with a different solution to generate docs from JSDocs we can move JSDocs improvement commits from here to master.

hekike commented:
[restify-docs-server-api.pdf](https://github.com/restify/node-restify/files/1308291/restify-docs-server-api.pdf)

Jekyll output from JSDoc generated markdown

yunong commented:
@DonutEspresso @retrohacker any more thoughts on this? My vote would be to land this since it's better than what we have so far.

retrohacker commented:
This is :fire:. I'm :+1: on landing this (after @hekike gets it to a point they are happy with) and reconciling the current docs with this approach moving forward. This will need an accompanying PR on https://github.com/restify/restify.github.io/blob/master/_data/docs.yml

hekike commented:
@DonutEspresso what do you think?

DonutEspresso commented on Makefile:
```diff
@@ -18,18 +18,22 @@
 # Files
 #
 DOC_FILES	 = index.restdown
+DOCUMENT_FILES = ./lib/index.js ./lib/server.js ./lib/plugins/audit.js ./lib/plugins/metrics.js ./lib/plugins/jsonp.js
```
What's the reason behind this whitelist?

hekike commented:
@DonutEspresso @yunong @retrohacker I finalized this PR, thanks for the feedback.

1. I skipped the prepush / precommit logic from this PR as it's not clear to me what's the best way as it would need a git add and commit after docs generation. It's a very similar challenge than CHANGELOG.md generation. For now, it's just a manual `make docs-build`

2. I changed the output for generated docs to `docs/_api`

3. Other than cleaning up the commit history I added a monster commit to this PR that reorders methods in the `server.js` file. It was necessary to have a better documentation output and cleaner code, I applied the following structure:

- public methods
  - constructor
  - server lifecycle methods (listen, close)
  - routing methods (get, head, post, patch, del, opts)
  - request lifecycle and middleware methods (pre, use, param, versionedUse, rm)
  - info and debug methods (address, inflightRequests, getDebugInfo, toString)
- private methods
- helper functions

If you are okay with these changes I think it's ready to :shipit:

retrohacker commented:
:shipit: :boat:
:tada:

DonutEspresso commented:
SGTM
